server/api/src/rsGenQuery.cpp

```diff
@@ -125,7 +125,84 @@ namespace {
         f << '\n';
     }
 
-}
+    auto get_resc_id_cond_for_hier_cond(const std::string_view _cond) -> std::string
+    {
+        // The default return string will yield 0 results when run in a query because RESC_ID is never '0'.
+        constexpr const char* default_condition_str = "='0'";
+
+        const std::string cond = _cond.data();
+
+        const auto open_quote_pos = cond.find_first_of("'");
+        const auto close_quote_pos = cond.find_last_of("'");
+        if (close_quote_pos == open_quote_pos) {
+            THROW(SYS_INVALID_INPUT_PARAM, fmt::format("Invalid condition: [{}]", cond));
+        }
+
+        const std::string op = cond.substr(0, open_quote_pos);
+
+        // Only allow =, !=, LIKE, and NOT LIKE to be specified. Check = first because it is used in some
+        // critical path operations and will only match 1 result, so it should be quick.
+        static const std::regex equal_regex{R"_(^\s*=\s*$)_"};
+        static const std::regex like_regex{R"_(^\s*(LIKE|like)\s*$)_"};
+        static const std::regex not_like_regex{R"_(^\s*(NOT|not)\s+(LIKE|like)\s*$)_"};
+        static const std::regex not_equal_regex{R"_(^\s*!=\s*$)_"};
+        bool equal_op = false;
+        bool like_op = false;
+        bool not_like_op = false;
+        bool not_equal_op = false;
+
+        // clang-format off
+        if (std::regex_match(op.data(), equal_regex))          equal_op = true;
+        else if (std::regex_match(op.data(), like_regex))      like_op = true;
+        else if (std::regex_match(op.data(), not_like_regex))  not_like_op = true;
+        else if (std::regex_match(op.data(), not_equal_regex)) not_equal_op = true;
+        // clang-format on
+
+        if (!equal_op && !like_op && !not_like_op && !not_equal_op) {
+            THROW(CAT_INVALID_ARGUMENT, fmt::format("Invalid condition: [{}]", cond));
+        }
+
+        // Generate a regex by replacing GenQuery wildcards with regex wildcards.
+        std::string hier_regex_str = cond.substr(open_quote_pos + 1, close_quote_pos - open_quote_pos - 1);
+        if (!equal_op && !not_equal_op) {
+            auto wildcard_pos = hier_regex_str.find_first_of("%");
+            while (std::string::npos != wildcard_pos) {
+                hier_regex_str.replace(wildcard_pos, 1, "(.*)");
+                wildcard_pos = hier_regex_str.find_first_of("%");
+            }
+        }
+        const std::regex hier_regex{hier_regex_str};
+
+        const auto want_to_match = like_op || equal_op;
+        std::vector<rodsLong_t> leaf_ids;
+        try {
+            for (const auto& hier : resc_mgr.get_all_resc_hierarchies()) {
+                const auto match_found = std::regex_match(hier, hier_regex);
+                if ((match_found && want_to_match) || (!match_found && !want_to_match)) {
+                    leaf_ids.push_back(resc_mgr.hier_to_leaf_id(hier));
+                }
+            }
```

Comment:
Consider adding a few comments explaining what is happening here.

L181 can be confusing on the first encounter.

Reply:
Will do